funding/commitment_type_negotiation.go

```diff
@@ -0,0 +1,120 @@
+package funding
+
+import (
+	"errors"
+
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// errUnsupportedExplicitNegotiation is an error returned when explicit
+	// channel commitment negotiation is attempted but either peer of the
+	// channel does not support it.
+	errUnsupportedExplicitNegotiation = errors.New("explicit channel " +
+		"type negotiation not supported")
+
+	// errUnsupportedCommitmentType is an error returned when a specific
+	// channel commitment type is being explicitly negotiated but either
+	// peer of the channel does not support it.
+	errUnsupportedChannelType = errors.New("requested channel type " +
+		"not supported")
+)
+
+// negotiateCommitmentType negotiates the commitment type of a newly opened
+// channel. If a channelType is provided, explicit negotiation for said type
+// will be attempted if the set of both local and remote features support it.
+// Otherwise, implicit negotiation will be attempted.
+func negotiateCommitmentType(channelType *lnwire.ChannelType,
+	local, remote *lnwire.FeatureVector) (lnwallet.CommitmentType, error) {
+
+	if channelType != nil {
+		if !hasFeatures(local, remote, lnwire.ExplicitChannelTypeOptional) {
+			return 0, errUnsupportedExplicitNegotiation
+		}
+		return explicitNegotiateCommitmentType(
+			*channelType, local, remote,
+		)
+	}
+
+	return implicitNegotiateCommitmentType(local, remote), nil
```

Comment:
Re the existing comment, if a user doesn't set this _new_ value when opening a channel, then we'll default to the old _implicit_ negotiation. IMO instead, we want to map the unknown value (so them not setting a value at all), to our current "default" channel type, and still use the explicit chan negotiation.